### DISCUSSION TITLE
How can I model a parallel machine scheduling problem as a VRP?

### DISCUSSION THREAD

**Discussion by abb-omidi** [PyVRP/PyVRP#571] [Q&A] [answered]
Dear support team,

I hope you are doing well. This is the first time I have worked with `PyVRP` library and I want to solve a special form of the scheduling problem. Let me describe more:

I want to solve a variant of the parallel machine scheduling problem in the presence of setup time and time frame/window. In the paper, this problem can be formulated as a VRPTW with some modifications. The tasks can be assumed as nodes also w.r.t a dummy node as a depot with its associated time and setup equal to zero. What actually I am looking for is how, if possible, I can define its coordinates matrix. At the moment, I have defined a coordinates matrix equal to zero and the demand equal to the processing time of each task, but I am unsure if it really being an appropriate way to define a coordinates matrix.

I was wondering if, have your insight regarding that and if possible to solve this problem with this library.

All the best

**Comment by N-Wouda** [comment]
Hi @abb-omidi! 👋 In order:

> In the paper, this problem can be formulated as a VRPTW with some modifications.

Which paper? Could you provide a link?

> What actually I am looking for is how, if possible, I can define its coordinates matrix.

What do you mean by 'coordinates matrix'? Is that your distance matrix? Or do you mean the depot and client coordinates?

**Comment by abb-omidi** [comment]
Dear @N-Wouda,

Thanks for your reply. I mean actually by **_in the paper_** was trying by hand an paper.  B.T.W. there are some references that makes this in my mind. E.g.:
* [Scheduling: Theory, Algorithms, and Systems, Fourth Edition by Michael L. Pinedo. P(82) in my edition.](https://link.springer.com/book/10.1007/978-1-4614-2361-4)
* [Memetic search for the min-max multiple traveling salesman problem with single and multiple depots](https://leria-info.univ-angers.fr/~jinkao.hao/papers/HeHaoEJOR2022.pdf)
* [Permutations Problems and how to solve them by Tony Hurlimann. P(4-5)](https://www.researchgate.net/publication/367265386_Permutations_Problems_and_how_to_solve_them)

Please, let me describe more about what I have tried to do.
I think a parallel machine scheduling problem can be assumed as a variant of VRP. In this case, the whole problem would be a graph such that the nodes define the tasks, the arcs define the processing procedure, and the vehicles define the machines. (e.g. $\{ \text{node1} \Leftrightarrow \text{node2} \}$ represents each node can be processed before the other if in the optimal solution, leads to a lower makespan.)

Also, there is a matrix something like a distance/duration matrix, which is usually named a **_setup matrix_**.  (In the scheduling problem defining a cost/distance matrix is somewhat weird). And already using the processing time of the tasks as a demand of the nodes. Indeed, uniform machines are equivalent to heterogenous fleet (maybe with a variable cost). But there are some differences as well.

Now, Based on the above description, I am unsure how I can define the coordinates matrix, as an input to `PyVRP`, as there exists no such a matrix in my problem.

I was wondering if, is it possible to use this library to solve my problem? (I think it should be possible, but I am not sure how)

Best regards
Abbas

**Reply by leonlan** [reply]
Hi @abb-omidi, it's possible to solve scheduling problems with PyVRP, but the scope is limited. For example, scheduling problems have very different objectives (minimizing makespan, total completion times, total tardiness, etc.), whereas routing mostly focuses on minimizing distance variants thereof. The equivalent of minimizing distance in scheduling would be minimizing setup times, which is a niche in the scheduling literature. Similarly, minimizing makespan is a niche in the routing literature.

I'll comment on parts of your question to help you a bit further, but I would suggest looking for another library to solve scheduling problems.

> I think a parallel machine scheduling problem can be assumed as a variant of VRP. In this case, the whole problem would be a graph such that the nodes define the tasks, the arcs define the processing procedure, and the vehicles define the machines. (e.g. 

This transformation is (almost) correct. An arc $(i, j)$ between two tasks is not a processing procedure but instead defines the setup time required when processing task $i$ after $j$.

> Also, there is a matrix something like a distance/duration matrix, which is usually named a setup matrix. (In the scheduling problem defining a cost/distance matrix is somewhat weird). And already using the processing time of the tasks as a demand of the nodes. Indeed, uniform machines are equivalent to heterogenous fleet (maybe with a variable cost). But there are some differences as well.

The processing time of a task can be transformed into the service time of a node, not the demand.
Because of this, uniform machines are not equivalent to a heterogeneous fleet. You need vehicle-specific client service times, but PyVRP does not support this since it's not that common in routing.

> Now, Based on the above description, I am unsure how I can define the coordinates matrix, as an input to PyVRP, as there exists no such a matrix in my problem.

It's not clear what you mean by coordinates matrix as there is no coordinates matrix in routing problems. Do you mean how to define coordinates for each location? In that case, you can simply use zero coordinates since there are no coordinates in a scheduling problem.

> I was wondering if, is it possible to use this library to solve my problem? (I think it should be possible, but I am not sure how)

Since you haven't really described a specific scheduling problem, it's hard to say if PyVRP can solve it. PyVRP can solve something along those lines:

- Identical parallel machines. That is, job-specific and not machine-specific processing times.
- Sequence-dependent setup times (the same for all machines)
	- This can become machine-specific when we add support #475.
- Objective: minimizing the sequence-dependent setup times.

**Reply by N-Wouda** [reply]
>Similarly, minimizing makespan is a niche in the routing literature.

I guess minimising makespan could be achieved by setting a uniform maximum route duration for each vehicle, and trying to lower that in an iterative fashion (keep lowering the maximum as long as feasible solutions are found by PyVRP). But that's indeed not really something I'd call "supported out of the box".

**Comment by abb-omidi** [comment]
Dear @leonlan,

Thank you so much for your detailed and very informative comments. 
By that actually I can get answers lots of my questions. We actually have a real-world class of optimization problem that fall in this ($P_m, s_1 \ | \ r_j, due_j, S_{i,j}, presd, Cap \ | \ C_{\max}/JIT$) category that unfortunately is NP-hard in an strong sense. We have tried some straightforward MIP or CP formulations. But they cannot go further beyond $80$% gap in MIP and $50$% in CP. 

Some follow up questions:
* Regarding the processing time and demand, how we can find the appropriate methods in `PyVRP` to write those?
* I am unsure I understand this part: `uniform machines are not equivalent to a heterogeneous fleet. You need vehicle-specific client service times, but PyVRP does not support this since it's not that common in routing.` Would you please, elaborate more on?
* As far as I see, my problem can be solved by your package based on the criteria you mentioned in the last part. Just for clarification, what do you mean by `(the same for all machines)`? Is this related to the tasks changeover/setup over the all machines and independent form machines?

I was wondering if, having your insight.
All the best

**Reply by leonlan** [reply]
> We have tried some straightforward MIP or CP formulations. But they cannot go further beyond % gap in MIP and % in CP.

Just to be sure you got this right: assuming that "gap" here refers to an optimality gap, 80% optimality gap in MIP or 50% optimality gap in CP doesn't mean that you have a bad solution. It may be that your primal solution is already (near-)optimal but you simply cannot find a better lower bound. The heuristics used in PyVRP do not provide a lower bound so there is no optimality gap at all.

> ($P_m, s_1 \ | \ r_j, due_j, S_{i,j}, presd, Cap \ | \ C_{\max}/JIT$

Assuming you follow the canonical scheduling classification scheme, I'm going to break down each part to get a better understanding of your problem and comment on whether it's possible to model this as VRP or not:
- $\alpha$: Machine environment
	- $P_m$: $m$ identical parallel machines
	- $s_1$: This is not standard so I'm not sure what this means. Is this a single machine that is used as an auxilliary resource when processing jobs on the parallel machines?
- $\beta$: Constraints
	- $r_j$: Release dates 
		- Can be modeled as earliest time windows in VRPs.
	- $due_j$: Due dates - I assume these are soft timing constraints.
		- This can be regarded as a soft time window in VRPs, but is not supported by PyVRP.
	- $S_{i,j}$: Setup times (machine-independent).
		- This is an arc duration in VRPs.
	- $presd$: I assume you mean $\text{prec}$ here, i.e., precedence constraints. 
		- Some variants of precedence constraints can be handled in VRPs, but the scheduling counterpart is often much more general. Anyway, precedence constraints are currently not supported in PyVRP and likely won't be until we implement #331.
	- $Cap$: This is non-standard but I assume this is capacity.
		- Is this a renewable or non-renewable resource? In case of the latter, then this is the same in VRPs. Renewable capacities are not supported in PyVRP.
- $\gamma$: Objective
	- $C_{\max}$: makespan.
		- This is not supported in PyVRP and likely won't ever be supported since it's very niche in VRPs.
	- JIT: This is any objective function that penalizes earliness and tardiness. 
		- I think this would be considered a "general time window/availability profile" in VRPs.
		- PyVRP does not support this and likely won't support this because it's again a very niche topic in VRPs.


> Regarding the processing time and demand, how we can find the appropriate methods in PyVRP to write those?

See the [examples](https://pyvrp.org/examples/quick_tutorial.html) and PyVRP documentation. You need the `service_duration` and `delivery` attribute in `Model.add_client`.

> As far as I see, my problem can be solved by your package based on the criteria you mentioned in the last part. Just for clarification, what do you mean by (the same for all machines)? Is this related to the tasks changeover/setup over the all machines and independent form machines?

Not really. Based on what we support right now we can solve: $P_m \mid r_{j}, s_{ij} \mid \sum s_{ij}$. 

Again, PyVRP is currently not the best way to solve your scheduling problem. You will have to build a custom solver to deal with your scheduling problem.

**Comment by abb-omidi** [comment]
Dear @leonlan,

Thanks once again for your explanations. 
> The heuristics used in PyVRP do not provide a lower bound so there is no optimality gap at all.

I knew the heuristics used in `PyVRP` did not provide a lower bound so there is no optimality. I managed to test some of instances of VRPs and see it makes very good solutions, and it is now enough for me. 

> Assuming you follow the canonical scheduling classification scheme ...

Your intuition regarding that is right. Also, let me say, I actually do not expect to solve my whole problem by this package. I have tried to decompose the original problem into sub-problems (also in a heuristic manner). Then using either exact or heuristic methods to solve these sub-problems. What I want to do by `PyVRP` is computing at least a good solution for a sub-problem. 

If the package can solve $P_m \mid (\text{general time window}), s_{ij} \mid \sum s_{ij}$, it would be great to try this and see how it performs again the real case data. **_(I will test and get back to you)_**.

> Again, PyVRP is currently not the best way to solve your scheduling problem. You will have to build a custom solver to deal with your scheduling problem.

Would you please, do you know any package to solve these kind of problems in mind? (I see you developed a package to solve job-shop problem. I have not tested it yet, but would you explain more regarding that?)

Regards

**Comment by leonlan** [comment]
> If the package can solve $P_m \mid (\text{general time window}), s_{ij} \mid \sum s_{ij}$, it would be great to try this and see how it performs again the real case data. **_(I will test and get back to you)_**.

PyVRP does not support general time windows.

> Would you please, do you know any package to solve these kind of problems in mind? (I see you developed a package to solve job-shop problem. I have not tested it yet, but would you explain more regarding that?)

I haven’t used any heuristic scheduling packages before. Some googling led to [pyscheduling](https://scheduling.cc/project/pyscheduling/), which looks like it can solve some variants of parallel machine scheduling problems. You could take their code to build your own solver.

The package PyJobShop is only a simple layer on top of a CP solver. Since you have already used CP, I would say there is not much use of that package solver-wise.

**Comment by abb-omidi** [comment]
Dear @leonlan,

Thank you so much for clarification regarding PyJobShop. (🙏)

Also, would you please, explain more about the implemented time window in PyVRP and its application?

Best

**Reply by leonlan** [reply]
PyVRP solves the standard version of VRPs with time windows. See, for example, the [documentation ](https://developers.google.com/optimization/routing/vrptw ) by Google OR-Tools to learn more about the VRPTW. You can also refer to the academic literature, which has a large number of studies on VRPTWs.

I don’t have the time to explain PyVRP’s implementation in detail. You can dig into the source code if you want to understand it better.

**Comment by leonlan** [comment]
@abb-omidi I just found a paper that reformulates uniform parallel machines as a multiple TSP, which is essentially a VRP. Preprint can be downloaded [here](https://riunet.upv.es/handle/10251/156847).

> Fanjul-Peyro, L., Ruiz, R., & Perea, F. (2019). Reformulations and an exact algorithm for unrelated parallel machine scheduling problems with setup times. Computers & Operations Research, 101, 173–182. https://doi.org/10.1016/j.cor.2018.07.007

The MILP formulation based on TSP is stronger than the scheduling MILP formulation. The paper may be useful for you in getting a stronger formulation for your problem.

**Comment by abb-omidi** [comment]
Dear @leonlan,

Thank you so much for pointing that out. Please, let me check somethings and I will get back to you. 

Regards

**Comment by abb-omidi** [comment]
Dear support team,

I hope you are doing well and I am sorry for my lateness.
I have tried to implement an instance of ($P_m \mid \text{standard time window}, s_{ij} \mid \sum s_{ij}$) based on my problem data.
What I am doing right now is defining a coordinate matrix, also assuming a dummy task as a depot point, with the values of zero. Then adding a service time as the processing time. I am unsure if I need to define a demand array, but I add something like [0,1,1...,1] just to ensure that each node should be visited at least once. (Please correct me if I am thinking wrong). 

In the next step, I add a duration matrix with a dummy task with zero value, based on $S_{i,j}$. Finally, adding a standard time frame/window. The corresponding code is:
```
COORDS = [
    (0, 0),  # location 0 - dummy task
    (0, 0),  # location 1
    (0, 0),  # location 2
    (0, 0),  # location 3
    (0, 0),  # location 4
    (0, 0),  # location 5
    (0, 0),  # location 6
    (0, 0),  # location 7
    (0, 0),  # location 8
    (0, 0),  # location 9
    (0, 0),  # location 10
]
DEMANDS = [0,1,1,1,1,1,1,1,1,1,1]
timings = [0,4,9,7,5,5,4,5,9,3,7]

DURATION_MATRIX = [
        [0,0,0,0,0,0,0,0,0,0,0],
        [0,0,5,3,2,2,1,2,5,0,3],
        [0,6,0,6,5,0,4,1,1,4,3],
        [0,2,2,0,1,4,5,1,4,5,2],
        [0,0,3,1,0,1,2,4,5,4,3],
        [0,2,0,2,0,0,1,4,3,5,2],
        [0,4,5,4,1,0,0,4,3,0,5],
        [0,0,1,3,5,1,0,0,4,2,2],
        [0,1,1,0,6,2,5,2,0,5,0],
        [0,1,0,1,3,1,1,2,2,0,6],
        [0,6,2,2,5,2,6,0,5,0,0]
]
TIME_WINDOWS = [
        (0, 999), 
        (61,70),
        (59,78),
        (1,67),
        (21,35),
        (6,17),
        (1,19),
        (57,68),
        (41,59),
        (1,12),
        (67,82)
]
```

```
m = Model()
m.add_vehicle_type(4, max_duration=100)
depot = m.add_depot(
    x=COORDS[0][0],
    y=COORDS[0][1],
    tw_early=TIME_WINDOWS[0][0],
    tw_late=TIME_WINDOWS[0][1],
)
clients = [
    m.add_client(
        x=COORDS[idx][0],
        y=COORDS[idx][1],
        tw_early=TIME_WINDOWS[idx][0],
        tw_late=TIME_WINDOWS[idx][1],
        service_duration=timings[idx]
    )
    for idx in range(1, len(COORDS))
]
```
By that, it seems the model works without any complaint, and the output is something like this:
```
PyVRP v0.8.2

Solving an instance with:
    1 depot
    10 clients
    4 vehicles (1 vehicle type)

                  |       Feasible        |      Infeasible
    Iters    Time |   #      Avg     Best |   #      Avg     Best
H     500      1s |  50        0        0 |  24       80        1
     1000      2s |  58        0        0 |  24       80        1
     1500      2s |  25        0        0 |  24       80        1
     2000      3s |  33        0        0 |  24       80        1

Search terminated in 3.00s after 2005 iterations.
Best-found solution has cost 0.

Solution results
================
    # routes: 4
   # clients: 10
   objective: 0.00
# iterations: 2005
    run-time: 3.00 seconds

Routes
------
Route #1: 1 7 10 
Route #2: 9 4 8 
Route #3: 3 6 5 
Route #4: 2 
```
Also, as long as I add the demand array the problem is infeasible. Would you say, is it a normal behavior?
I was wondering if, I am in the right way or need to change/adjust something? 

All the best

**Reply by leonlan** [reply]
There is no need to define demands for your problem.

**Comment by abb-omidi** [comment]
Dear @leonlan,

Thank you for clarifying. It seems I stand in the right way. Just a follow-up question:
* How is the objective function interpreted? (I mean, in the optimal solution, e.g. for the first assignment, it starts from dummy tasks, goes to task $1$, then task $7$, task $10$, and finally gets again back to the dummy task. If the objective is to minimize $\sum s_{i,j}$ , the corresponding setup time would be: (task_0 $\rightarrow$ task_1: $0$, task_1 $\rightarrow$ task_7: $2$, task_7 $\rightarrow$ task_10: $2$, task_10 $\rightarrow$ task_0: $0$). That seems not to be equal to zero. 

Thanks in advance

**Reply by N-Wouda** [reply]
See [here](https://pyvrp.github.io/v0.8.1/api/pyvrp.html#pyvrp._pyvrp.CostEvaluator.cost). This will change a bit in 0.9.0.

**Reply by leonlan** [reply, marked as answer]
The costs are defined by the distances of edges, but you have only set edge durations. If you set the distance attribute in the ‘add_edge’ method then you should get the right objective.

As Niels mentioned above, starting from v0.9.0, it will be possible to have an objective based on durations but for v0.8.0 you need to specify distances.

**Comment by abb-omidi** [comment]
Dear @N-Wouda, @leonlan,

Thank you so much for your time and advice in converting my problem to what the `PyVRP` package can solve.
Also, I will try something new and if I have any questions, I will get back to you.

Best regards

**Comment by abb-omidi** [comment]
Dear @leonlan,

Would you please, 

* Is it possible to use `PyVRP` v.0.9.0 on the Google-Colab? (It seems the version installed there is `pyvrp-0.8.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl`).
* Is there any knowledge regarding the approximation ratio of the algorithm used in `PyVRP` rather than the optimal solution? (e.g. 1 + $\epsilon$)

All the best

**Reply by N-Wouda** [reply]
> * Is it possible to use `PyVRP` v.0.9.0 on the Google-Colab? (It seems the version installed there is `pyvrp-0.8.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl`).

Can Google Colab install from source? v0.9.0 is not yet released, so the best you could get for now is to install directly from the source materials in this repository.

> * Is there any knowledge regarding the approximation ratio of the algorithm used in `PyVRP` rather than the optimal solution? (e.g. 1 + ϵ)

Not that I am aware of, since we do not use any particular techniques that have known approximation ratios.

**Comment by abb-omidi** [comment]
Dear @N-Wouda,

Many thanks for your comments. 
1) At the moment I do not know if it may be installed from the source, but I will check that.

2) > Not that I am aware of, since we do not use any particular techniques that have known approximation ratios.

Thanks for clarifying that.

Best regards
